src/pykeen/nn/init.py

```diff
@@ -375,21 +375,20 @@ def __init__(
         # create random walk matrix
         adj = torch_ppr.utils.prepare_page_rank_adjacency(edge_index=edge_index, num_nodes=num_entities)
         num_entities = adj.shape[0]
-        # allocate tensor
-        tensor = torch.zeros(num_entities, dim)
-        # note: create tensor to extract diagonals
-        # torch.diag does not work with sparse tensors
-        diag_indices = torch.arange(num_entities).unsqueeze(0).repeat(2, 1)
-        one_diag = torch.sparse_coo_tensor(indices=diag_indices, values=torch.ones(num_entities))
-        # iterate
-        matrix = adj
-        for i in range(dim):
-            matrix = torch.sparse.mm(matrix, adj)
-            # extract diagonal; torch.diag does not work with sparse tensors
-            diag = (matrix * one_diag).coalesce()
-            indices = diag.indices()
-            # we need to use indices here, since there may be zero diagonal entries
-            tensor[indices, i] = diag.values() * ((i + 1) ** (space_dim / 2))
+
+        tensor = torch.stack(
+            [
+                (i ** (space_dim / 2)) * diag
+                for i, diag in enumerate(
+                    map(
+                        functools.partial(extract_diagonal_sparse_or_dense, eye=sparse_eye(n=num_entities)),
+                        iter_matrix_power(matrix=adj, max_iter=dim),
+                    ),
+                    start=1,
+                )
+            ],
+            dim=-1,
```

Comment:
having re-written this in the `map` fashion, can we introduce some sort of caching to matrix powers? 
now when we need to extract, eg 3-d vectors, we'd need to perform 1 + 2 + 3 = 6 sparse mms instead of previous 3 - this will affect large graphs of FB15k237 size

Reply:
It does already cache, since `iter_matrix_power` does all the heavy lifting 🏋️

Reply:
tbh, while the clarity of the individual steps did improve quite a bit, this map part lost of a bit of clarity compared to the loop we had there before 😅

Reply:
agree, I liked an explicit computation loop better

Reply:
01e194d - I tried to improve clarify here again